tools/run_tests/xds_k8s_test_driver/framework/xds_url_map_testcase.py

```diff
@@ -0,0 +1,590 @@
+# Copyright 2021 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A test framework built for urlMap related xDS test cases."""
+
+import abc
+import unittest
+import json
+import time
+import inspect
+import threading
+import functools
+from typing import Any, Union, Tuple, Mapping, Iterable, Optional
+
+from framework import xds_flags
+from framework import xds_k8s_flags
+from framework.infrastructure import gcp
+from framework.infrastructure import k8s
+from framework.infrastructure import traffic_director
+from framework.test_app import client_app
+from framework.test_app import server_app
+
+from absl import flags
+from absl import logging
+from absl.testing import absltest
+from google.protobuf import json_format
+import googleapiclient
+from kubernetes.utils.create_from_yaml import FailToCreateError
+
+logging.set_verbosity(logging.DEBUG)
+flags.adopt_module_key_flags(xds_flags)
+flags.adopt_module_key_flags(xds_k8s_flags)
+
+_STRATEGY = flags.DEFINE_enum('strategy',
+                              default='create',
+                              enum_values=['create', 'keep', 'reuse'],
+                              help='Strategy of GCP resources management')
+# TODO(lidiz) find a better way to filter test cases
```

Comment:
see `-k` flag https://docs.python.org/3/library/unittest.html#cmdoption-unittest-k

Reply:
A large part of testing behavior is controlled by flags, but `unittest` command itself doesn't allow command line arguments. And `absltest` is built for Bazel, many functionalities, like test loading, test filtering, are controlled by Bazel flags. If we instrument these tests with Bazel, we will get them for free.

Technically, we can hack the arguments, like hard code `args` or build a portable from environment var to absl flags. That doesn't feel ideal.

Please let me know if you think we should support the `unittest` command.

```
usage: python3 -m unittest [-h] [-v] [-q] [--locals] [-f] [-c] [-b] [-k TESTNAMEPATTERNS] [tests ...]

positional arguments:
  tests                a list of any number of test modules, classes and test methods.

optional arguments:
  -h, --help           show this help message and exit
  -v, --verbose        Verbose output
  -q, --quiet          Quiet output
  --locals             Show local variables in tracebacks
  -f, --failfast       Stop on first fail or error
  -c, --catch          Catch Ctrl-C and display results so far
  -b, --buffer         Buffer stdout and stderr during tests
  -k TESTNAMEPATTERNS  Only run tests which match the given substring

Examples:
  python3 -m unittest test_module               - run tests from test_module
  python3 -m unittest module.TestClass          - run tests from module.TestClass
  python3 -m unittest module.Class.test_method  - run specified test method
  python3 -m unittest path/to/test_file.py      - run tests from test_file.py

usage: python3 -m unittest discover [-h] [-v] [-q] [--locals] [-f] [-c] [-b] [-k TESTNAMEPATTERNS] [-s START] [-p PATTERN] [-t TOP]

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         Verbose output
  -q, --quiet           Quiet output
  --locals              Show local variables in tracebacks
  -f, --failfast        Stop on first fail or error
  -c, --catch           Catch Ctrl-C and display results so far
  -b, --buffer          Buffer stdout and stderr during tests
  -k TESTNAMEPATTERNS   Only run tests which match the given substring
  -s START, --start-directory START
                        Directory to start discovery ('.' default)
  -p PATTERN, --pattern PATTERN
                        Pattern to match tests ('test*.py' default)
  -t TOP, --top-level-directory TOP
                        Top level directory of project (defaults to start directory)
```